spec.bs

```diff
@@ -6,57 +6,417 @@ Status: CG-DRAFT
+
+Because bits of entropy can leak via {{SharedStorage/selectURL()}} if a user activates the fenced frame that navigates to the resulting opaque URL, we impose on each origin a <dfn>bit budget</dfn>, i.e. a maximum allowance of entropy bits that are permitted to leak during a given [=epoch=], where we define an <dfn>epoch</dfn> as a time period whose duration is a fixed pretermined length. When a leak occurs, its value in entropy bits is calculated and stored as a <dfn>bit debit</dfn> with a timestamp, and only [=bit debit=]s whose timestamps are within the current [=epoch=] are used to determine an origin's remaining budget. When an origin has insufficient remaining budget, {{SharedStorage/selectURL()}} will return its default value.
+
+<xmp class='idl'>
+  [Exposed=(Window,SharedStorageWorklet)]
+  interface SharedStorage {
+    [Exposed=(Window)] 
+    Promise<any> run(DOMString name,
+                     optional SharedStorageRunOperationMethodOptions options = {});
+
+    [Exposed=(Window)] 
+    Promise<USVString> selectURL(DOMString name,
                                  FrozenArray<SharedStorageUrlWithMetadata> urls,
                                  optional SharedStorageRunOperationMethodOptions options = {});
 
-  Promise<any> run(DOMString name, optional SharedStorageRunOperationMethodOptions options = {});
+    Promise<any> set(DOMString key,
+                     DOMString value,
+                     optional SharedStorageSetMethodOptions options = {});
+
+    Promise<any> append(DOMString key,
+                        DOMString value);
+
+    Promise<any> delete(DOMString key);
+
+    Promise<any> clear();
+
+    [Exposed=(SharedStorageWorklet)] Promise<DOMString> get(DOMString key);
+
+    [Exposed=(SharedStorageWorklet)] Promise<unsigned long> length();
+
+    [Exposed=(SharedStorageWorklet)] Promise<double> remainingBudget();
 
-  readonly attribute SharedStorageWorklet worklet;
-};
+    [Exposed=(SharedStorageWorklet)] SharedStorageIterator keys();
 
-[Exposed=(Window)]
-interface SharedStorageWorklet {
-  Promise<undefined> addModule(USVString moduleURL);
-};
+    [Exposed=(SharedStorageWorklet)] SharedStorageIterator entries();
 
-dictionary SharedStorageSetMethodOptions {
-  boolean ignoreIfPresent;
-};
+    [Exposed=(Window)] readonly attribute SharedStorageWorklet worklet;
+  };
 
-dictionary SharedStorageRunOperationMethodOptions {
-  object data;
-};
+  dictionary SharedStorageSetMethodOptions {
+    boolean ignoreIfPresent;
+  };
 
-dictionary SharedStorageUrlWithMetadata {
-  required USVString url;
-  object reportingMetadata;
-};
+  dictionary SharedStorageRunOperationMethodOptions {
+    object data;
+  };
+
+  enum IteratorMode { "key", "keyValue" };
+
+  [Exposed=(SharedStorageWorklet)]
+  interface SharedStorageIterator {
+    constructor(IteratorMode mode);
+    Promise<object> next();
+  };
 </xmp>
+
+  ## Note on Parallelism ## {#parallelism}
```

Comment:
The worklet spec already has the notion of **agent** for a worklet: e.g. https://html.spec.whatwg.org/multipage/worklets.html#create-a-worklet-global-scope. It’s roughly a "thread". Perhaps we still need to define more about the process/thread model (i.e. the implementation defined part), but at least it’s better to relate to the existing terms, so that we don’t need to define a separate task queue for the worklet operations. e.g. for run() and selectURL(), after the initial sanity checks, we can:
1. Let outsideSettings be worklet's relevant settings object.
2. Let agent be the result of obtaining a worklet agent given outsideSettings. Run the rest of these steps in that agent.
	- Let |operationMap| be this SharedStorageWorkletGlobalScope's operation map ...

To ensure the ordering of DB operations, we can define a parallel queue just for the DB. e.g. <dfn for="browsing context">shared storage database parallel queue</dfn>. And mention that it is associated with each browsing context and that the worklet scope will be associated with the same queue. And whenever you need to access the DB, "Enqueue the following steps on |context|'s [=browsing context/shared storage database parallel queue=]".

Reply:
Done.